### ISSUE TITLE
init() should be called delayed

### ISSUE THREAD

**Issue by oliverhausler** [DreaminginCodeZH/MaterialProgressBar#12]
MaterialProgressBar fully initializes even though it is set to INVISIBLE or GONE. It should instead only call init() when it first becomes visible.

I did some profiling and could improve overall time to set up a view to 20% just by removing MaterialProgressBar. I admit I used it very often, but you should still initialize lazy right before it is needed.

**Comment by zhanghai** [comment]
Inflating the framework XML vector drawable should be much more expansive than creating the custom Java drawable in this library, except the creation of compat animators and interpolators on platforms prior to Lollipop where I workarounded by generating some float array for `PropertyValuesHolder` with `PathMeasure` (see [ObjectAnimatorCompatBase](library/src/main/java/me/zhanghai/android/materialprogressbar/internal/ObjectAnimatorCompatBase.java) and the source of support-v4 `PathInterpolatorCompat`).

It would be nice if you could investigate the bottleneck of the performance problem, whether this only happens on old platforms, and where is critical (for example I doubt whether I should have interpolated 500 points for accuracy now).

**Comment by oliverhausler** [comment]
I forgot to say that, sorry. It happens on old platforms <5.0.

**Comment by zhanghai** [comment]
In other words, framework also loads drawables at creation instead of upon becoming visible, so I would like to pinpoint the issue first. (And perhaps your suggested solution will lag the animation during usage when the app animates a `MaterialProgressBar` from GONE to VISIBLE.)

**Comment by oliverhausler** [comment]
Yes, I tried to set it to "gone" and it was still loaded, which might be a framework flaw.

I have a ScrollView in my app with hundreds of Views. Each view also had a MaterialProgressBar which I had set to gone, and which I only made visible when I needed it.

The view has like 10 embedded views (all built-in except MaterialProgressBar). As loading the ScrollView was slow (6 secs to populate 400 views), I investigated. Removing the built-in views made no difference, but removing MaterialProgress bar did (1.5 secs for 400 views).

I don't know what the framework exactly does, but it works very well to keep several hundred views in a ScrollView and only populate them when they become visible, so it seems the framework behaves correctly with built-in views.

**Comment by zhanghai** [comment]
BTW, I've just checked that support-v4 `FastOutSlowInInterpolator` sampled 201 points, half the number of this library. Don't know whether this can be a performance hit.

Another way around is to pre-sample the points as magic numbers so that no runtime calculation is needed - but this also loses the direct dependency on the orignal path, so as I did not encounter a performance problem myself and partially considered pre-5.0 devices as only supporting I did not choose that solution.

**Comment by oliverhausler** [comment]
I understand. Everything is a big mess under Android <50 and if it works well on Android 5.0+, it's a reasonable decision not to make optimizations for older versions.

**Comment by zhanghai** [comment]
You mean 400 views on a pre-5.0 device? wow. Yes, in that situation this can really be a performance hit as each animator did not share their interpolated data as I only expected two or three `ProgressBar`s inside one activity.

And did you test on a Lollipop device? What is the situation there?

As a personal advice, I would strongly recommend that you switch to a recyclable view container as `ListView`, `GridView` or even better `RecyclerView`, because this is a huge performance anti-pattern  compared to the optimization I left behind in this library. And even if I did the static optimization, the 1.5 second should still be too long a lag for user because it must block your UI thread.

**Comment by oliverhausler** [comment]
I haven't tested on a 5.0+ device and I don't have one here, sorry. But it won't give us a reasonable comparison either, probably, as it will be faster.

I tried implementing GridView, but it had by far worse performance than what I wrote. Problem is that users fling when scrolling, and recycling views takes much too long. So what I do is I place empty layouts into ScrollView and populate the views inside only when they get close to or within the viewport. The only problem that I have with that approach is that animator within ScrollView keeps frames which require memory, just because of the size of the view area. Otherwise performance is great. I guess for the next version we will need to write a mix between RecyclerView and what I currently do.

Something else: I didn't find a method in your library to set circular or horizontal progress bar from code. Is this missing? I get circular by default.

**Comment by zhanghai** [comment]
I'm willing to implement the optimization if it is outstanding enough and not much too dirty, and if you switched to recyclable containers and still find it annoying.

The planned optimizations are:
- [x] Cut down on `NUM_POINTS` inside `ObjectAnimatorCompatBase` from `500` to `201` as in support-v4.
- [ ] Use a `WeakHashMap` and `WeakReference`/`SoftReference` for cache the results since the path arguments passed in are static.
- [ ] Or stop using `ObjectAnimatorCompatBase` and resort to pre-sampled animators.

**Comment by zhanghai** [comment]
Yes, the setter is just missing as the framework did not give you this - the size of `ProgressBar` and the drawable is determined by some `minWidth` and `maxWidth` magic (see its `onMeasure()`), and I did not find a good way to change that programmtically. So it is not easy for you to reuse a `ProgressBar` both horizontally and circularly. This is also the reason why you still need a style from this library set upon the target XML instead of being automatically calculated.

In other words to control the apperance of `ProgressBar` it seems that framework forced you to go the XML way.

**Comment by oliverhausler** [comment]
I didn't try to switch, but I tried instantiating progressBar from code, only when I need it, instead of loading it from XML.

`progressBar = new MaterialProgressBar(getContext());` works, but it is circular by default. When I try to do `new MaterialProgressBar(getContext(), null, R.style.CustomProgress)` [not sure if I can leave attrs = null] it crashes:

12-12 18:52:18.171 13423-13423/cc.closeup.android E/AndroidRuntime: FATAL EXCEPTION: main
 Process: cc.closeup.android, PID: 13423
 java.lang.UnsupportedOperationException: Determinate circular drawable is not yet supported
 at me.zhanghai.android.materialprogressbar.MaterialProgressBar.init(MaterialProgressBar.java:89)
 at me.zhanghai.android.materialprogressbar.MaterialProgressBar.(MaterialProgressBar.java:51)
 at cc.closeup.android.views.ChannelPreview.setProgressVisible(ChannelPreview.java:423)

**Comment by oliverhausler** [comment]
oh, and something else that I also didn't find: I could set the background color from code, but not the foreground color. I tried to change the color while it's running [I am abusing it as progress bar for live video, which changes colors from grey to color dependent on the connect status].

**Comment by oliverhausler** [comment]
"When I try to do new MaterialProgressBar(getContext(), null, R.style.CustomProgress) [not sure if I can leave attrs = null] it crashes:"

Style looks like this BTW:

``` xml
    <style name="CustomProgress" parent="Widget.MaterialProgressBar.ProgressBar.Horizontal">
        <item name="android:minHeight">@dimen/progressbar_height</item>
        <item name="android:maxHeight">@dimen/progressbar_height</item>
    </style>
```

**Comment by zhanghai** [comment]
First of all you need an attribute (instead of style) resource (R.attr.*) for the third parameter, see `android.content.res.Resources.Theme#obtainStyledAttributes(AttributeSet, int[], int, int)` which says:

> defStyleAttr An attribute in the current theme that contains a reference to a style resource that supplies defaults values for the TypedArray.  Can be 0 to not look for defaults.

That exception is thrown if you have `app:mpb_progressStyle="circular"`  && (`android:indeterminate="false"` || `app:mpb_setBothDrawables="true"`).

Your style looks good, though. (**EDIT: Just ensure that you've added attributes I mentioned above.**)

And for the foreground color issue, I guess you are looking for `setTint()`, which I've backported to 4.x.

---

In order to not disturb watchers with issue not related to this library I'm putting it here.

> recycling views takes much too long.

Generally it should not take too long because you should just be setting different text and images on them. If image loading is your problem, consider using another layer of in-memory cache (such as Glide). If having a number of diverse layouts is your problem, try telling the container you have different types of views to recycle. Sorry if I told you something you already knew, I'm just thinking that it should not generally be a problem and reinventing the wheels is error-prone and seldomly benefical.

**Comment by oliverhausler** [comment]
Something of your markdown got lost in your previous post. Probably because you had `<>` in there which Github understands wrong.

**Comment by zhanghai** [comment]
Please refresh the page - Chrome on Linux is giving me intermittent font rendering artifacts so sometimes I have to reload the page to discover something wrong.

**Comment by oliverhausler** [comment]
now I see it, thanks

**Comment by oliverhausler** [comment]
@DreaminginCodeZH Thanks for your help!

**Comment by zhanghai** [comment]
You are welcome. Did you solve your problem? Do you still think that the performance needs to be optimized?

**Comment by oliverhausler** [comment]
I went to bed cause it was late here. I'll get back to you in case I can't make it work.

Regarding performance, if it works on Android 5.0+ I would leave it. Not sure if we do further tests or just instantiate from code. I'll discuss today.

**Comment by zhanghai** [comment]
Closing due to inactivity.
